src/test/java/bsh/InterpreterTest.java

```diff
@@ -52,4 +55,63 @@ public void check_system_object() throws Exception {
         assertEquals(Boolean.FALSE, TestUtil.eval("return bsh.system.shutdownOnExit;"));
     }
 
+    @Test
+    public void interrupt_on_tokenizing() throws Exception {
+        final StringReader in = new StringReader("print(\"\n");
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+        Interpreter bsh = new Interpreter(in, new PrintStream(baos), System.err, false);
+        bsh.setShutdownOnExit(false);
+
+        Thread t = new Thread(bsh);
+        t.start();
+
+        Thread.sleep(250); // give Interpreter eom time to set itself up
+
+        //
+        // now the interpreter should be on a blocking read; we interrupt it,
+        // give it some time to make the magic and make sure it tells us about it
+        //
+        bsh.interrupt(); Thread.sleep(250);
+
+        assertTrue(baos.toString().contains("... interrupted ..."));
+        assertTrue(!baos.toString().contains("Parser Error:"));
+
+        //
+        // we should also be able to issue new statements, but this would require
+        // some more refactoring, let's defer it to the next iteration
+        //
+    }
+
+    @Test
+    public void interrupt_on_parsing() throws Exception {
+        final StringReader in = new StringReader("class A {\n"); // generate a ParseException
+        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        final ByteArrayOutputStream err = new ByteArrayOutputStream();
+
```

Comment:
I hope this noisy code shows the need of some more advanced TDD tools... creating BAOS is not fun! :D

Reply:
I am open to suggestions just needs to be motivated. I don't think there are many tests currently using this  method so shouldn't be a major issue.

Reply:
The suggestion is using junit system-rules (https://stefanbirkner.github.io/system-rules/). the major issue is that there are too few test ;), and making them easier to write will help.